corehq/apps/app_manager/suite_xml.py

```diff
@@ -1966,42 +2005,33 @@ def get_target_module(case_type, module_id, with_product_details=False):
             #    will be loading the same case type
             # see advanced_app_features#child-modules in docs
             datum_pairs = list(izip_longest(datums, root_datums))
+            datum_ids = {d['datum'].id: d for d in datums}
             index = 0
             changed_ids_by_case_tag = {}
             for this_datum_meta, parent_datum_meta in datum_pairs:
                 if not this_datum_meta:
                     continue
-
                 this_datum = this_datum_meta['datum']
                 action = this_datum_meta['action']
-                if action and action.parent_tag in changed_ids_by_case_tag:
-                    # update any reference to previously changed datums
-                    change = changed_ids_by_case_tag[action.parent_tag]
-                    nodeset = this_datum.nodeset
-                    old = session_var(change['old_id'])
-                    new = session_var(change['new_id'])
-                    this_datum.nodeset = nodeset.replace(old, new)
-
+                update_refs(action, this_datum, changed_ids_by_case_tag)
                 if not parent_datum_meta:
                     continue
 
-                that_datum = parent_datum_meta['datum']
-                if this_datum.id != that_datum.id:
+                parent_datum = parent_datum_meta['datum']
+                if this_datum.id != parent_datum.id:
                     if not parent_datum_meta['requires_selection']:
                         datums.insert(index, parent_datum_meta)
                     elif this_datum_meta['case_type'] == parent_datum_meta['case_type']:
-                        action = action
-                        if action:
-                            changed_ids_by_case_tag[action.case_tag] = {
-                                "old_id": this_datum.id,
-                                "new_id": that_datum.id
-                            }
-                        this_datum.id = that_datum.id
+                        avoid_duplicate_ids(action, this_datum, parent_datum, datum_ids, changed_ids_by_case_tag)
+                    elif (
+                        module.module_type == 'basic' and
+                        module.parent_select.active and
+                        module.parent_select.module_id == root_module.unique_id
+                    ):
+                        datums.insert(index, parent_datum_meta)
```

Comment:
Hmmm! When I was stepping through the code originally I must have thought that parent_datum_meta['requires_selection'] might be true for basic modules. But I must have been wrong, because it's working correctly without that `elif` for the example I was using. I'll take it out.